internal/resources/providers/msgraph/provider.go

```
@@ -0,0 +1,91 @@
+}
+
+type provider struct {
+	log    *logp.Logger
+	client interface {
+		ServicePrincipals() *serviceprincipals.ServicePrincipalsRequestBuilder
+	}
+}
+
+// Second argument is scopes. Leave nil, then it selects default; Adjust if in trouble
+// Docs: https://learn.microsoft.com/en-us/graph/sdks/create-client?from=snippets&tabs=go
+func NewProvider(log *logp.Logger, azureConfig auth.AzureFactoryConfig) (ProviderAPI, error) {
+	// Requires 'Directory.Read.All' API permission
+	c, err := graph.NewGraphServiceClientWithCredentials(azureConfig.Credentials, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating MS Graph client: %w", err)
+	}
+
+	p := &provider{
+		log:    log.Named("msgraph"),
+		client: c,
+	}
+
+	return p, nil
+}
+
+// Docs:
+// - https://github.com/microsoftgraph/msgraph-sdk-go
+// - https://learn.microsoft.com/en-us/graph/api/serviceprincipal-list?view=graph-rest-beta&tabs=go
+// - https://learn.microsoft.com/en-us/graph/sdks/paging?tabs=go
+func (p *provider) ListServicePrincipals(ctx context.Context) ([]*models.ServicePrincipal, error) {
+	requestConfig := &serviceprincipals.ServicePrincipalsRequestBuilderGetRequestConfiguration{}
+
+	response, err := p.client.ServicePrincipals().Get(ctx, requestConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error listing Azure Service Principals: %w", err)
+	}
+
+	pageIterator, err := graphcore.NewPageIterator[*models.ServicePrincipal](
+		response,
+		p.client.ServicePrincipals().RequestAdapter,
+		models.CreateServicePrincipalCollectionResponseFromDiscriminatorValue,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error paging Azure Service Principals: %w", err)
+	}
+
+	items := []*models.ServicePrincipal{}
+	err = pageIterator.Iterate(ctx, func(pageItem *models.ServicePrincipal) bool {
+		items = append(items, pageItem)
+		return true // to continue the iteration
+	})
+	if err != nil {
+		p.log.Errorf("error iterating over Service Principals: %v", err)
```

shouldn't we return the error if principals are not returned?

---

This is a log, because we might still get some resources. The pagination failed, but `items` slice has been populated. We try to get the best scenario for us - getting data, even if incomplete.